### PR TITLE
Quadlet: Add support for .build files

### PR DESCRIPTION
**Pull request by jmaibaum**
#### Does this PR introduce a user-facing change?

```release-note
Quadlet: Add support for .build files
```

This adds support for `.build` Quadlet units, as per discussion in https://github.com/containers/podman/discussions/17043#discussioncomment-9347329, basically by mapping Key/Value pairs inside a `.build` file to `podman build` command line arguments.

You depend on a container image built by a `.build` file, by adding `Image=name.build` to your `.container` file, similar to how you can depend on an `.image` file.

Mandatory keys in a `.build` file under the `[Build]` group are so far:

```
[Build]
# Tag the image you intend to build so that container service files can use this name in their command lines 
ImageTag=localhost/imagename
# Either of
File=/path/to/a/Containerfile
# or
SetWorkingDirectory=unit
# to add a build context to `podman build`
```

Not every single `podman build` command line argument is being handled, yet. There might be more which would be nice to have before merging (please let me know if you think I missed an important one!), but since adding a new Key/Value pair is a rather cumbersome and boring process, I decided to stop after adding the following:

- all `KeyXYZ` I found already defined, and which I also found in `podman builds` supported arguments (notable exceptions from this rule are those I did never use before myself, or those where I found more complicated parsing code in the current quadlet code base)
- a few mandatory things which were missing so far, like `KeyFile` (`podman build`'s `--file`, i.e. the Containerfile path), or `KeyTarget`

I have added e2e tests, and documentation for all currently supported keys/args, closely following what I found for the already existing Quadlet file types.

This is my first contribution to a Go project, so please let me know in case I did not follow best practice or used unidiomatic constructs. I used `go fmt` on all Go files I touched (except for `quadlet_test.go`, as I only added single lines here, and wasn't sure if `go fmt` would alter parts of the file I did not touch), and generally tried to follow naming conventions I found in surrounding code.

I have tested this with a few toy projects locally, and there it did it's job. Apart from adding more command line args if deemed necessary, I consider the implementation ready functionality wise.

Happy to hear what the maintainers think about this!

### CODE REVIEW

**Comment by dustymabe** [comment]
As the person who started the discussion in https://github.com/containers/podman/discussions/17043 I'd like to say thank you for working on this!

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:52]: It seems we have a cyclic dependency problem between `.volume` and `.build` files. A `.volume` can depend on a `.build` file for it's `Image` key while a `.build` file can depend on `.volume` file for its `Volume` key. I'm not talking about the same files (that's the user's problem), but on different couples of files.

I had a similar issue when I added the support for `.pod` files. The way I solved it was by doing an initial run on `.pod` files (See `generatePodsInfoMap`) and passing it around.

File: pkg/systemd/quadlet/quadlet.go
Comment (ygalblum) [inline pkg/systemd/quadlet/quadlet.go:332]: There are three additional keys all Quadlet files support. I know the correct implementation should handle it in one place, but, it currently doesn't. The keys are:
- `ContainersConfModule` and `GlobalArgs` which are handled by calling `createBasePodmanCommand`. These flags control arguments that are at podman level, not specific to the command
- `PodmanArgs` which is handled by calling `handlePodmanArgs`. This key allows passing arguments that do not have a dedicated Quadlet key 

I see that you are using these functions correctly, but you are not allowing their usage

**Review by ygalblum — COMMENTED** [review]
Thanks for working on this. Great job.

The dependency comment is obviously the tricky one

**Comment by ygalblum** [comment]
BTW the build is failing on trailing white spaces in the doc file

File: cmd/quadlet/main.go
Comment (jmaibaum) [reply cmd/quadlet/main.go:52]: Good catch! Yes, this seems like a tricky issue. So I think I need to do two additional runs similar to how you've handled a pod's containers:

1. Check all .volumes for referenced .build files in the Image Key
2. Check all .build files for referenced .volume files in a Volume Key, and here also bail out in case the user added a cyclic dependency

File: pkg/systemd/quadlet/quadlet.go
Comment (jmaibaum) [reply pkg/systemd/quadlet/quadlet.go:332]: Thanks for this hint. Indeed, I didn't know that these exist and I have to add them to the supported keys manually.

File: cmd/quadlet/main.go
Comment (ygalblum) [reply cmd/quadlet/main.go:52]: We try not to have Quadlet catch user errors. So, I don't think you need to handle any bailouts.
I think a simpler solution would be to calculate the `resourceNames` map in a preliminary run and then do the full conversion. At the end of the day you just need the names, Quadlet does not run anything

File: pkg/systemd/quadlet/quadlet.go
Comment (jmaibaum) [reply pkg/systemd/quadlet/quadlet.go:332]: I have fixed this in my latest push.

**Comment by jmaibaum** [comment]
> BTW the build is failing on trailing white spaces in the doc file

I have fixed this in my latest push.

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:622]: Just make sure to remove these prints :)

File: docs/source/markdown/podman-systemd.unit.5.md
Comment (ygalblum) [inline docs/source/markdown/podman-systemd.unit.5.md:1344]: Please also copy the documentation of the additional fields: `PodmanArgs`, `ContainersConfModule` and `GlobalArgs` to both the table and the details below it

**Review by ygalblum — COMMENTED** [review]
Just these two small comments

File: cmd/quadlet/main.go
Comment (ygalblum) [reply cmd/quadlet/main.go:52]: Yes, I agree that such a change should have its own PR

File: docs/source/markdown/podman-systemd.unit.5.md
Comment (jmaibaum) [reply docs/source/markdown/podman-systemd.unit.5.md]: Done.

File: cmd/quadlet/main.go
Comment (jmaibaum) [reply cmd/quadlet/main.go:52]: Thanks. I will thus add the missing test cases next.

**Comment by jmaibaum** [comment]
Yes, I am still working on the tests, currently struggling with the ".volume requires .build/.image" tests. I am going to push 2 failing tests shortly, where I do not understand why they fail, as they work correctly when I am running quadlet on these files manually.

File: test/e2e/quadlet_test.go
Comment (jmaibaum) [inline test/e2e/quadlet_test.go:864]: This test (and the similar one for the `.image` file) is currently failing, even though it works when I am manually running `quadlet` on the test file. The error message during test is:

```
> Enter [It] Volume - Quadlet image (.build) - /var/home/jm/go/src/github.com/containers/podman/test/e2e/quadlet_test.go:864 @ 05/16/24 15:26:59.847
  Running: /var/home/jm/go/src/github.com/containers/podman/bin/quadlet --user --no-kmsg-log /tmp/podman-e2e-1875717062/subtest-2494887151/generated with QUADLET_UNIT_DIRS=/tmp/podman-e2e-1875717062/subtest-2494887151/quadlet
  quadlet-generator[122417]: converting "build.quadlet.volume": requested Quadlet image basic.build was not found
  [FAILED] Failure recorded during attempt 1:
  Expected
      <int>: 1
  to match exit code:
      <int>: 0
```

Yet, `basic.build` and `basic.image` are existing.

Comment (jmaibaum) [reply test/e2e/quadlet_test.go:864]: Is it possible to keep the temp directories the test suite creates?

**Comment by jmaibaum** [comment]
OK, I found the issue with the 2 testcases `image.quadlet.volume` and `build.quadlet.volume`. The reason is that `handleImageSource()` really requires the image names to be found in the resourceNames array. For all other `*.quadlet.*` tests requiring other quadlet files the `handleXYZ()` functions fall back to the "default name", like `systemd-$name` for networks or volumes.

The two tests succeed when I patch `quadlet_test.go` to always copy `basic.image` and `basic.build` alongside the file to be tested.

Since `handleImageSource` needs the `*.image` or `*.build` files to read the image name from though, I guess we need a special test case type where you can specify additionally required files. Unless you see an easier way, I am going to write this test type function next.

My assumption is that this complication with `handleImageSource()` was the reason why the

```
[Volume]
Driver=image
Image=some.image
```

hasn't been tested before...

File: test/e2e/quadlet_test.go
Comment (jmaibaum) [reply test/e2e/quadlet_test.go]: I have fixed this issue by adding a new testcase type: "Running quadlet test case with dependencies"

**Comment by jmaibaum** [comment]
After reading the changelog for Podman 5.1.0-rc1 I also added the `--group-add` support to Build units.

Unless you have further remarks or suggestions @ygalblum , I consider the PR ready. Please let me know and then I will squash the commits into one,

**Comment by rhatdan** [comment]
Please fix the man pages so the tests will run.

**Comment by jmaibaum** [comment]
The failing RPM builds seem to be due to infra issues unrelated to the changes in this PR IIUC?

**Review by ygalblum — APPROVED** [review]
Thanks for all your effort.

File: docs/source/markdown/podman-systemd.unit.5.md
Comment (jmaibaum) [reply docs/source/markdown/podman-systemd.unit.5.md:1352]: Thanks for finding this. Will fix this when I am back (currently only on mobile).

File: docs/source/markdown/podman-systemd.unit.5.md
Comment (jmaibaum) [reply docs/source/markdown/podman-systemd.unit.5.md:1513]: I guess we can pass any string different from `unit` or `file` directly to `podman build` That should make it possible to pass a generic URL.

Comment (jmaibaum) [reply docs/source/markdown/podman-systemd.unit.5.md:1513]: I took a look at this, it's not that easy because of a bunch of corner cases. I am working on it, but it will take likely a few days.

**Comment by jmaibaum** [comment]
@ygalblum I have a WIP for supporting URLs, and also for handling more corner cases in terms of relative or absolute file paths to Containerfiles (in both `File=` and `SetWorkingDirectory=`). It is not yet ready (still fighting with a few of those corner cases), but I think it makes sense to wait with merging until these useful features (as mentioned in the code conversation above) can be handled correctly.

IIUC, this PR would be targeting a release with Podman 5.2.0 earliest, as the 5.1.0-rc period has already started. So, I would like to take this time to polish this.

File: docs/source/markdown/podman-systemd.unit.5.md
Comment (dustymabe) [reply docs/source/markdown/podman-systemd.unit.5.md:1513]: Thank you!

I think this is a really big feature that will be very useful. I use something like this already today (where I use manually crafted systemd units) where I'll push to git and then just reboot the server and it will build and then run a container.

Today I use .container files for the container but I have to implement the build stuff myself. This will make things so much easier.

Comment (jmaibaum) [reply docs/source/markdown/podman-systemd.unit.5.md:1516]: I have just pushed an initial take on supporting more of `podman build`'s options to specifying `--file`s or build context. URLs should be supported now. Would be great if someone could test this on a few of their use cases, as well. I, for myself, had not been using the URL support of `podman build` so far, so I am unsure if I have already covered all cases.

**Comment by jmaibaum** [comment]
@ygalblum Initial support for URLs pushed. I did this in a separate commit on purpose to be able to revisit it again in case you (or another maintainer) don't like it. I will squash again if this is acceptable.

Also, I would be really glad if someone could give this a test spin. As noted above, I had not been using the URL features of `podman build` before myself. More eyes (and test runs) welcome!

File: pkg/systemd/quadlet/quadlet.go
Comment (SEIAROTg) [inline pkg/systemd/quadlet/quadlet.go:1403]: I'd suggest we use the same logic in `podman build` [here][1], which handles more types of URL.

[1]: https://github.com/containers/podman/blob/34d2d20d46c1ad3cf13432fc481047f1b285ff20/cmd/podman/common/build.go#L206.

File: test/e2e/quadlet/file-abs.build
Comment (ygalblum) [inline test/e2e/quadlet/file-abs.build:1]: Why not use just `assert-podman-args`?

Comment (jmaibaum) [reply test/e2e/quadlet/file-abs.build:1]: I used `final-args`, because this has `--file` as the last argument instead of a context directory.

File: pkg/systemd/quadlet/quadlet.go
Comment (jmaibaum) [reply pkg/systemd/quadlet/quadlet.go]: I have to take a closer look at this if it's just adding a bunch more checks for different URL types (`git://` def. makes sense). I didn't do this yet, because the postman build docs explicitly only mention http(s) for `--file`. But I am curious why people can't just put a https:// in front of their GitHub URLs? Why the special treatment of GitHub which just doesn't look like a URL, and thus could be misdetected as a relative path?

Comment (jmaibaum) [reply pkg/systemd/quadlet/quadlet.go]: For now I have expanded the regex check I added to include github.com/ and git:// links which seems to work according to the e2e tests. Please let me know if you want me to follow the `podman build` code even more closely.

Comment (SEIAROTg) [reply pkg/systemd/quadlet/quadlet.go]: Looks good to me!

**Comment by jmaibaum** [comment]
I tried to get the system tests working locally, but failed both for `--rootless` and `--root`, most likely because I am running them inside https://containertoolbx.org/ on Fedora Silverblue 40 (i.e. it's podman inside podman).

The `--rootless` tests fail due to some issue with uid maps:

```
$ ./hack/bats --rootless 252:"quadlet - basic"
--------------------------------------------------
$ bats --filter quadlet - basic test/system/252-quadlet.bats
252-quadlet.bats
 ✗ quadlet - basic
   [teardown] $ podman pod rm -t 0 --all --force --ignore
   time="2024-05-27T16:25:54+02:00" level=error msg="running `/usr/bin/newuidmap 7575 0 1000 1 1 100000 65536`: newuidmap: write to uid_map failed: Operation not permitted\n" time="2024-05-27T16:25:54+02:00" level=error msg="invalid internal status, try resetting the pause process with \"/var/home/jm/go/src/github.com/containers/podman/bin/podman system migrate\": cannot set up namespace using \"/usr/bin/newuidmap\": exit status 1"
   [teardown] $ podman rm -t 0 --all --force --ignore
   time="2024-05-27T16:25:54+02:00" level=error msg="running `/usr/bin/newuidmap 7593 0 1000 1 1 100000 65536`: newuidmap: write to uid_map failed: Operation not permitted\n" time="2024-05-27T16:25:54+02:00" level=error msg="invalid internal status, try resetting the pause process with \"/var/home/jm/go/src/github.com/containers/podman/bin/podman system migrate\": cannot set up namespace using \"/usr/bin/newuidmap\": exit status 1"
   [teardown] $ podman network prune --force
   time="2024-05-27T16:25:54+02:00" level=error msg="running `/usr/bin/newuidmap 7611 0 1000 1 1 100000 65536`: newuidmap: write to uid_map failed: Operation not permitted\n" time="2024-05-27T16:25:54+02:00" level=error msg="invalid internal status, try resetting the pause process with \"/var/home/jm/go/src/github.com/containers/podman/bin/podman system migrate\": cannot set up namespace using \"/usr/bin/newuidmap\": exit status 1"
   [teardown] $ podman volume rm -a -f
   time="2024-05-27T16:25:54+02:00" level=error msg="running `/usr/bin/newuidmap 7630 0 1000 1 1 100000 65536`: newuidmap: write to uid_map failed: Operation not permitted\n" time="2024-05-27T16:25:54+02:00" level=error msg="invalid internal status, try resetting the pause process with \"/var/home/jm/go/src/github.com/containers/podman/bin/podman system migrate\": cannot set up namespace using \"/usr/bin/newuidmap\": exit status 1"
   (from function `bail-now' in file test/system/helpers.bash, line 227,
    from function `die' in file test/system/helpers.bash, line 813,
    from function `run_podman' in file test/system/helpers.bash, line 423,
    from function `basic_setup' in file test/system/helpers.bash, line 142,
    from function `setup' in test file test/system/252-quadlet.bats, line 27)
     `basic_setup' failed
   
   [16:25:54.037398006] $ /var/home/jm/go/src/github.com/containers/podman/bin/podman rm -t 0 --all --force --ignore
   [16:25:54.175595621] time="2024-05-27T16:25:54+02:00" level=error msg="running `/usr/bin/newuidmap 7534 0 1000 1 1 100000 65536`: newuidmap: write to uid_map failed: Operation not permitted\n"
   time="2024-05-27T16:25:54+02:00" level=error msg="invalid internal status, try resetting the pause process with \"/var/home/jm/go/src/github.com/containers/podman/bin/podman system migrate\": cannot set up namespace using \"/usr/bin/newuidmap\": exit status 1"
```

The `--root` tests fail due to write permissions:

```
$ sudo ./hack/bats --root 252:"quadlet - basic"
# bats --filter quadlet - basic test/system/252-quadlet.bats
252-quadlet.bats
 ✗ [252] quadlet - basic
   (from function `bail-now' in file test/system/helpers.bash, line 235,
    from function `assert' in file test/system/helpers.bash, line 929,
    from function `run_quadlet' in file test/system/252-quadlet.bats, line 64,
    in test file test/system/252-quadlet.bats, line 164)
     `run_quadlet "$quadlet_file"' failed
   
   [16:26:38.037388720] # /var/home/jm/go/src/github.com/containers/podman/bin/podman rm -t 0 --all --force --ignore
   
   [16:26:38.182784656] # /var/home/jm/go/src/github.com/containers/podman/bin/podman ps --all --external --format {{.ID}} {{.Names}}
   
   [16:26:38.311321046] # /var/home/jm/go/src/github.com/containers/podman/bin/podman images --all --format {{.Repository}}:{{.Tag}} {{.ID}}
   [16:26:38.427485648] quay.io/libpod/testimage:20240123 1f6acd4c4a1d
   # /var/home/jm/go/src/github.com/containers/podman/bin/quadlet  /run/systemd/system
   quadlet-generator[8061]: generating service file /run/systemd/system/basic_by6ktASTep.service: open /run/systemd/system/basic_by6ktASTep.service: permission denied
```

So, I'm afraid I can't add system tests, because I cannot meaningfully test them locally. If they are required before this PR can be merged, I will need help fixing any of the above errors.

**Comment by ygalblum** [comment]
@jmaibaum I think we can do with the system tests for now. I've restarted the failed tests (which seem unrelated to this change)

**Comment by ygalblum** [comment]
@containers/podman-maintainers PTAL

**Comment by rhatdan** [comment]
/lgtm
Great work @jmaibaum

**Pull request merged**
